File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
class Solution {
    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];

        // 행렬을 초기화 합니다.
        int[][] matrix = new int[rows+1][columns+1];
        int cnt = 1;
        for(int i = 1 ; i <= rows ; i++){
            for(int j = 1 ; j <= columns ; j++){
                matrix[i][j] = cnt;
                cnt += 1;
            }
        }

        // 쿼리를 하나씩 확인합니다
        cnt = 0;
        for(int[] query: queries){
            // 하나의 쿼리에서 시계방향으로 회전합니다. 이때 최솟값도 저장합니다.
            int x1 = query[0], y1 = query[1], x2 = query[2], y2 = query[3];
            
            int minValue = Integer.MAX_VALUE;
            int last = matrix[x1][y1];
            // x1 (y1 ~ y2)로 이동
            for(int i = y1; i < y2; i++){
                int cur = matrix[x1][i+1];
                minValue = Math.min(minValue,cur);
                matrix[x1][i+1] = last;
                last = cur;
            }

            // y2 (x1 ~ x2)로 이동
            for(int i = x1; i < x2 ; i++){
                int cur = matrix[i+1][y2];
                minValue = Math.min(minValue, cur);
                matrix[i+1][y2] = last;
                last = cur;
            }

            // x2 (y2 ~ y1)으로 이동
            for(int i = y2; i > y1; i--){
                int cur = matrix[x2][i-1];
                minValue = Math.min(minValue, cur);
                matrix[x2][i-1] = last;
                last = cur;
            }

            // y1 (x2 ~ x1)으로 이동
            for(int i = x2; i > x1 ; i--){
                int cur = matrix[i-1][y1];
                minValue = Math.min(minValue, cur);
                matrix[i-1][y1] = last;
                last = cur;
            }

            answer[cnt] = minValue;
            cnt += 1;
        }
        

        return answer;
    }
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.kt
import kotlin.math.min

class Solution {
    fun solution(rows: Int, columns: Int, queries: Array<IntArray>): IntArray {
    var answer = mutableListOf<Int>()
    val MAX = 100 * 200

    // 행렬 초기화
    val matrix = (0 until rows).map { r ->
        (1 .. columns).map { c ->
            r * columns + c
        }.toIntArray()
    }.toTypedArray()

    // 쿼리 돌아가면서 회전시킨다
    for(query in queries){
        var minValue = MAX
        val (x1, y1, x2, y2) = query.map{it -1}

        val queue = ArrayDeque<Int>()
        // 오른쪽
        for(yy in y1..y2){
            queue.add(matrix[x1][yy])
            minValue = min(matrix[x1][yy], minValue)
        }

        // 아래
        for(xx in x1+1..x2){
            queue.add(matrix[xx][y2])
            minValue = min(matrix[xx][y2], minValue)
        }

        // 왼쪽
        for(yy in y2-1 downTo y1){
            queue.add(matrix[x2][yy])
            minValue = min(matrix[x2][yy], minValue)
        }


        // 위로
        for(xx in x2-1 downTo x1){
            queue.add(matrix[xx][y1])
            minValue = min(matrix[xx][y1], minValue)
        }


        val last = queue.removeLast()
        queue.addFirst(last)

        // 오른쪽
        for(yy in y1..y2){
            matrix[x1][yy] = queue.removeFirst()
            // println("$x1,$yy : ${matrix[x1][yy]}")
        }

        // 아래
        for(xx in x1+1..x2){
            matrix[xx][y2] = queue.removeFirst()
            // println("$xx,$y2 : ${matrix[xx][y2]}")
        }

        // 왼쪽
        for(yy in y2-1 downTo y1){
            matrix[x2][yy] = queue.removeFirst()
            // println("$x2,$yy : ${matrix[x2][yy]}")
        }


        // 위로
        for(xx in x2-1 downTo x1){
            matrix[xx][y1] = queue.removeFirst()
            // println("$xx,$y1 : ${matrix[xx][y1]}")
        }

        answer.add(minValue)
    }

    // 할 때 마다 최솟값 업데이트 시킨다

    return answer.toIntArray()
    }
}